File: py/fastapi-pypika-example/src/api/ninja/ninja.py
from uuid import UUID
from fastapi import Depends, Response, APIRouter
from ...db.providers import get_providers, Providers
from ...types.types import NinjaNew, NinjaUpdates

router = APIRouter()


@router.get("/{id}")
async def get_ninja(
    response: Response,
    id: str,
    providers: Providers = Depends(get_providers)
):
    """
    Get ninja
    """
    if not isinstance(id, str):
        response.status_code = 400
        return {}
    try:
        ninja = await providers.pgdal.get_ninja(UUID(id))
        response.status_code = 200
        return ninja
    except Exception:
        response.status_code = 500
        return {}


@router.post("/")
async def create_ninja(
    response: Response,
    ninja_new: NinjaNew,
    providers: Providers = Depends(get_providers)
):
    """
    Insert new ninja
    """
    try:
        ninja = await providers.pgdal.create_ninja(ninja_new)
        response.status_code = 200
        return ninja
    except Exception:
        response.status_code = 500
        return {}


@router.put("/{id}")
async def update_ninja(
    response: Response,
    id: str,
    ninja_updates: NinjaUpdates,
    providers: Providers = Depends(get_providers)
):
    """
    Update existing ninja
    """
    if not isinstance(id, str):
        response.status_code = 400
        return {}
    try:
        ninja = await providers.pgdal.update_ninja(UUID(id), ninja_updates)
        response.status_code = 200
        return ninja
    except Exception:
        response.status_code = 500
        return {}


@router.delete("/{id}")
async def delete_ninja(
    response: Response,
    id: str,
    providers: Providers = Depends(get_providers)
):
    """
    Delete existing ninja
    """
    if not isinstance(id, str):
        response.status_code = 400
        return {}
    try:
        ninja = await providers.pgdal.delete_ninja(UUID(id))
        response.status_code = 200
        return ninja
    except Exception:
        response.status_code = 500
        return {}


@router.post("/{ninja_id}/jutsu/{jutsu_id}")
async def associate_ninja_and_jutsu(
    response: Response,
    ninja_id: str,
    jutsu_id: str,
    providers: Providers = Depends(get_providers)
):
    """
    Associate ninja and jutsu
    """
    if not isinstance(ninja_id, str) and isinstance(jutsu_id, str):
        response.status_code = 400
        return {}
    try:
        await providers.pgdal.associate_ninja_and_jutsu(UUID(ninja_id), UUID(jutsu_id))
        response.status_code = 200
        return {}
    except Exception:
        response.status_code = 500
        return {}


@router.delete("/{ninja_id}/jutsu/{jutsu_id}")
async def dissociate_ninja_and_jutsu(
    response: Response,
    ninja_id: str,
    jutsu_id: str,
    providers: Providers = Depends(get_providers)
):
    """
    Dissociate ninja and jutsu
    """
    if not isinstance(ninja_id, str) and isinstance(jutsu_id, str):
        response.status_code = 400
        return {}
    try:
        await providers.pgdal.dissociate_ninja_and_jutsu(UUID(ninja_id), UUID(jutsu_id))
        response.status_code = 200
        return {}
    except Exception:
        response.status_code = 500
        return {}


@router.get("/{id}/jutsus")
async def get_ninja_with_jutsus(
    response: Response,
    id: str,
    providers: Providers = Depends(get_providers)
):
    """
    Get ninja with jutsus
    """
    if not isinstance(id, str):
        response.status_code = 400
        return {}
    try:
        ninja = await providers.pgdal.get_ninja_with_jutsus(UUID(id))
        response.status_code = 200
        return ninja
    except Exception:
        response.status_code = 500
        return {}

File: go/chi-squirrel-example/api/ninja/ninja.go
package ninja

import (
	"net/http"

	chi "github.com/go-chi/chi/v5"

	"chi-squirrel-example/api/utils"
	"chi-squirrel-example/providers"
	"chi-squirrel-example/types"
)

func RegisterNinja(app chi.Router, providers *providers.Providers) {
	app.Route("/api/ninja", func(r chi.Router) {
		// Get ninja
		r.Get("/{id}/", func(w http.ResponseWriter, r *http.Request) {
			id := chi.URLParam(r, "id")
			ninja, err := providers.PGDAL.GetNinja(id)
			if err != nil {
				utils.SendStatus(w, http.StatusInternalServerError)
				return
			}
			utils.SendJSON(w, http.StatusOK, ninja)
		})

		// Insert ninja
		r.Post("/", func(w http.ResponseWriter, r *http.Request) {
			var ninjaNew types.NinjaNew
			err := utils.ReadJSON(r, &ninjaNew)
			if err != nil {
				utils.SendStatus(w, http.StatusBadRequest)
				return
			}
			ninja, err := providers.PGDAL.CreateNinja(ninjaNew)
			if err != nil {
				utils.SendStatus(w, http.StatusInternalServerError)
				return
			}
			utils.SendJSON(w, http.StatusOK, ninja)
		})

		// Update ninja
		r.Put("/{id}/", func(w http.ResponseWriter, r *http.Request) {
			id := chi.URLParam(r, "id")
			var ninjaUpdates types.NinjaNew
			err := utils.ReadJSON(r, &ninjaUpdates)
			if err != nil {
				utils.SendStatus(w, http.StatusBadRequest)
				return
			}
			ninja, err := providers.PGDAL.UpdateNinja(id, ninjaUpdates)
			if err != nil {
				utils.SendStatus(w, http.StatusInternalServerError)
				return
			}
			utils.SendJSON(w, http.StatusOK, ninja)
		})

		// Delete Ninja
		r.Delete("/{id}/", func(w http.ResponseWriter, r *http.Request) {
			id := chi.URLParam(r, "id")
			ninja, err := providers.PGDAL.DeleteNinja(id)
			if err != nil {
				utils.SendStatus(w, http.StatusInternalServerError)
				return
			}
			utils.SendJSON(w, http.StatusOK, ninja)
		})

		// Get ninja with jutsus
		r.Get("/{id}/jutsus/", func(w http.ResponseWriter, r *http.Request) {
			id := chi.URLParam(r, "id")
			ninja, err := providers.PGDAL.GetNinjaWithJutsus(id)
			if err != nil {
				utils.SendStatus(w, http.StatusInternalServerError)
				return
			}
			utils.SendJSON(w, http.StatusOK, ninja)
		})
	})
}

func RegisterNinjaJutsu(app chi.Router, providers *providers.Providers) {
	app.Route("/api/ninja-jutsu/", func(r chi.Router) {
		// Associate ninja & jutsu
		r.Post("/{ninjaID}/{jutsuID}/", func(w http.ResponseWriter, r *http.Request) {
			ninjaID := chi.URLParam(r, "ninjaID")
			jutsuID := chi.URLParam(r, "jutsuID")
			err := providers.PGDAL.AssociateNinjaJutsu(ninjaID, jutsuID)
			if err != nil {
				utils.SendStatus(w, http.StatusInternalServerError)
				return
			}
			utils.SendStatus(w, http.StatusOK)
		})

		// Dissociate ninja & jutsu
		r.Delete("/{ninjaID}/{jutsuID}/", func(w http.ResponseWriter, r *http.Request) {
			ninjaID := chi.URLParam(r, "ninjaID")
			jutsuID := chi.URLParam(r, "jutsuID")
			err := providers.PGDAL.DissociateNinjaJutsu(ninjaID, jutsuID)
			if err != nil {
				utils.SendStatus(w, http.StatusInternalServerError)
				return
			}
			utils.SendStatus(w, http.StatusOK)
		})
	})
}
